File: Binary Search/Python/Minimized Maximum of Products Distributed to Any Store.py
# Problem link: https://leetcode.com/problems/minimized-maximum-of-products-distributed-to-any-store/description/
# Time: O(n log max(quantities))
# Space: O(1)
class Solution:
    def minimizedMaximum(self, n: int, quantities: List[int]) -> int:
        result, left, right = -1, 1, max(quantities)
        

        def isPossible(diff):
            total = 0

            for q in quantities:
                total += ceil(q / diff)
            
            return total <= n

        while left <= right:
            mid = (left + right) // 2

            if(isPossible(mid)):
                result = mid
                right = mid - 1
            else:
                left = mid + 1
        
        return result

File: Binary Search/Java/Minimized Maximum of Products Distributed to Any Store.java
// Problem link: https://leetcode.com/problems/minimized-maximum-of-products-distributed-to-any-store/description/
// Time: O(n log max(quantities))
// Space: O(1)
class Solution {
    public int minimizedMaximum(int n, int[] quantities) {
        int left = 1, right = 0, result = -1, mid;

        for(int i : quantities)
            right = Math.max(right, i);

        while(left <= right) {
            mid = (left + right) / 2;

            if(isPossible(quantities, n, mid)) {
                result = mid;
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }

        return result;
    }
    private boolean isPossible(int[] quantities, int n, int diff) {
        int total = 0;

        for(int q : quantities)
            total += Math.ceil((double) q / diff);

        return total <= n;
    }
}
